src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/AsyncCausalityTracer.cs

```diff
@@ -0,0 +1,99 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading.Tasks
+{
+    internal static class AsyncCausalityTracer
+    {
+        internal static void Enable()
+        {
+            Interlocked.Increment(ref listenerCnt);
+        }
+
+        internal static void Disable()
+        {
+            Interlocked.Decrement(ref listenerCnt);
+        }
+
+        internal static bool LoggingOn => listenerCnt > 0;
+        private static int listenerCnt = 0;
+
+        // The TraceXXX methods should be called only if LoggingOn property returned true
+        //
+        [MethodImpl(MethodImplOptions.NoInlining)] // Tracking is slow path. Disable inlining for it.
+        internal static void TraceOperationCreation(Task task, string operationName)
+        {
+            try
+            {
+                TplEventSource.Log.TraceOperationBegin(task.Id, operationName, RelatedContext: 0);
+            }
+            catch (Exception ex)
+            {
+                LogAndDisable(ex);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        internal static void TraceOperationCompletion(Task task, AsyncCausalityStatus status)
+        {
+            try
+            {
+                TplEventSource.Log.TraceOperationEnd(task.Id, status);
+            }
+            catch (Exception ex)
+            {
+                LogAndDisable(ex);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        internal static void TraceOperationRelation(Task task, CausalityRelation relation)
+        {
+            try
+            {
+                TplEventSource.Log.TraceOperationRelation(task.Id, relation);
+            }
+            catch (Exception ex)
+            {
+                LogAndDisable(ex);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        internal static void TraceSynchronousWorkStart(Task task, CausalitySynchronousWork work)
+        {
+            try
+            {
+                TplEventSource.Log.TraceSynchronousWorkBegin(task.Id, work);
+            }
+            catch (Exception ex)
+            {
+                LogAndDisable(ex);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        internal static void TraceSynchronousWorkCompletion(CausalitySynchronousWork work)
+        {
+            try
+            {
+                TplEventSource.Log.TraceSynchronousWorkEnd(work);
+            }
+            catch (Exception ex)
+            {
+                LogAndDisable(ex);
+            }
+        }
+
+        // we should catch and log exceptions but never propagate them.
+        private static void LogAndDisable(Exception ex)
+        {
+            Disable();
```

Comment:
I'm not clear why this is the right thing to do.  If an exception occurs (and per my previous comment, I'm not sure why/when that would happen), we decrement the ref count, but we don't actually unregister anything.  Doesn't that suggest that the count could just keep decreasing and decreasing?  That would seem to a) mean you couldn't re-enable it eventually, and b) in an extreme case, it could wrap around and we'd end up seeing it as enabled again?

Reply:
An exception can be thrown when TplEventSource tries to write events into ETW and an error occurs (i.e. not enough memory). That being said, I'm not entirely sure why we should catch that exception and disable... Like Jan said, it is a remnant from the old WinRT AsyncCausalityTracer. I suppose we can remove this unless @noahfalk has any other thoughts?

Reply:
I think it would be fine to remove these exception catches. I don't see any reason why an exception that was acceptable to throw from TplEventSource would be unacceptable to throw here.